transport/src/main/java/io/netty/bootstrap/DynamicLocalSocketAddress.java

```diff
@@ -0,0 +1,123 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.bootstrap;
+
+import io.netty.util.internal.ObjectUtil;
+
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.NetworkInterface;
+import java.net.SocketAddress;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.Iterator;
+
+/**
+ * {@link SocketAddress} which will provide the local {@link SocketAddress} to bind to after the remote address
+ * is resolved and so can be used with {@link Bootstrap#connect(SocketAddress, SocketAddress)}.
+ */
+public abstract class DynamicLocalSocketAddress extends SocketAddress {
+
+    /**
+     * Return a {@link SocketAddress} to bind to for the given {@code remote} {@link SocketAddress} or {@code null}
+     * if no specific should be used.
+     */
+    public abstract SocketAddress address(SocketAddress remote);
+
+    public static DynamicLocalSocketAddress networkInterface(NetworkInterface networkInterface, int port) {
+        return new DynamicInetSocketAddress(networkInterface, port);
+    }
+
+    public static DynamicLocalSocketAddress any(Iterable<? extends InetSocketAddress> iterable) {
+        return new DynamicInetSocketAddress(iterable);
+    }
+
+    public static DynamicLocalSocketAddress either(InetSocketAddress inet4Address, InetSocketAddress inet6Address) {
+        if (!(inet4Address.getAddress() instanceof Inet4Address)) {
+            throw new IllegalArgumentException();
+        }
+        if (!(inet6Address.getAddress() instanceof Inet6Address)) {
+            throw new IllegalArgumentException();
+        }
+
+        return new DynamicInetSocketAddress(Arrays.asList(inet4Address, inet6Address));
+    }
+
+    private static final class DynamicInetSocketAddress extends DynamicLocalSocketAddress {
+        private final Iterable<? extends InetSocketAddress> iterable;
+
+        DynamicInetSocketAddress(final NetworkInterface networkInterface, final int port) {
+            this(new Iterable<InetSocketAddress>() {
+                @Override
+                public Iterator<InetSocketAddress> iterator() {
+                    return new EnumerationIterator(networkInterface.getInetAddresses(), port);
+                }
+            });
+            ObjectUtil.checkNotNull(networkInterface, "networkInterface");
+        }
+
+        DynamicInetSocketAddress(Iterable<? extends InetSocketAddress> iterable) {
+            this.iterable = ObjectUtil.checkNotNull(iterable, "iterable");
+        }
+
+        @Override
+        public SocketAddress address(SocketAddress remote) {
+            if (!(remote instanceof InetSocketAddress)) {
+                throw new IllegalArgumentException("Only InetSocketAddress is supported");
```

Comment:
How about appending `remote` to the message for better debuggability?

Reply:
Remove the implementation